pennylane/templates/subroutines/qpe.py

```diff
@@ -0,0 +1,68 @@
+# Copyright 2018-2021 Xanadu Quantum Technologies Inc.
+
+# Licensed under the Apache License, Version 2.0 (the "License");
+# you may not use this file except in compliance with the License.
+# You may obtain a copy of the License at
+
+#     http://www.apache.org/licenses/LICENSE-2.0
+
+# Unless required by applicable law or agreed to in writing, software
+# distributed under the License is distributed on an "AS IS" BASIS,
+# WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+# See the License for the specific language governing permissions and
+# limitations under the License.
+"""
+Contains the ``QuantumPhaseEstimation`` template.
+"""
+from numpy.linalg import matrix_power
+
+import pennylane as qml
+from pennylane.templates.decorator import template
+from pennylane.wires import Wires
+
+
+@template
+def QuantumPhaseEstimation(unitary, target_wires, estimation_wires):
+    r"""Performs the
+    `quantum phase estimation <https://en.wikipedia.org/wiki/Quantum_phase_estimation_algorithm>`__
+    circuit.
+
+    Given a unitary :math:`U`, this template applies the circuit for quantum phase
+    estimation. The unitary is applied to the qubits specified by ``target_wires`` and :math:`n`
+    qubits are used for phase estimation as specified by ``estimation_wires``.
+
+    .. figure:: ../../_static/templates/subroutines/qpe.svg
+        :align: center
+        :width: 60%
+        :target: javascript:void(0);
+
+    Args:
+        unitary (array): the phase estimation unitary
+        target_wires (Union[Wires, Sequence[int], or int]): the target wires to apply the unitary
+        estimation_wires (Union[Wires, Sequence[int], or int]): the wires to be used for phase
+            estimation
+
+    Raises:
+        QuantumFunctionError: if the ``target_wires`` and ``estimation_wires`` share a common
+            element
+
+    .. UsageDetails::
+
+        TODO
+    """
+
+    target_wires = Wires(target_wires)
+    estimation_wires = Wires(estimation_wires)
+
+    if len(Wires.shared_wires([target_wires, estimation_wires])) != 0:
+        raise qml.QuantumFunctionError("The target wires and estimation wires must be different")
+
+    for i, wire in enumerate(estimation_wires):
+        qml.Hadamard(wire)
+
+        # Could we calculate the matrix power more efficiently by diagonalizing?
```

Comment:
Good question! I think the answer is yes, it will be faster to diagonalize the unitary first. Here's my back-of-the-envelope calculation:

Matrix multiplication scales as `O(n^3)` for matrices of dimension `n`.Each new power is one more matrix multiplication, so we need to perform at least `2^k` multiplications. This gives a runtime of `O(2^k n^3)`.

Diagonalizing a matrix also takes time `O(n^3)`, which only needs to be done once. Then we need to take the powers of each of the `n` eigenvalues `2^k` times, but each power is `O(1)` since these are scalars, so the whole thing takes time `O(n2^k)`. Once that's done, we need to perform another set of multiplications to transform the unitaries back to non-diagonal form. This takes time `O(k n^3)`. So the total runtime is ~`O(n^3) + O(n 2^k) +O(k n^3)` which is smaller than `O(2^k n^3).` Potentially much smaller if `k` is large.

So yes, I'd suggest performing the diagonalization here

Reply:
Thanks JM! Makes sense for the diagonalizing matrix complexity.

I think doing it like this: 1f700a1da2de258d2d7e3f9051b2abd4b0fb6c2a
also has a `O((k - 1) * n^3)` complexity.

i.e., we need `u, u^2, u^4, ... u^(2^(k - 1))`.
Going from `u` to `u^2` is one matrix multiplication, going from `u^2` to `u^4` is another matrix multiplication, overall requiring `k-1` matrix multiplications? I guess the trick is we don't need all of the matrix powers of `u`, just the ones that correspond to a power of 2?